## Imports

In [1]:
import pandas as pd
import numpy as np

### Check Numpy Version

In [2]:
np.__version__

'1.23.2'

The version is 1.23.2

### Load Dataset

In [3]:
df = pd.read_csv("data.csv")

# Displaying first 10 rows
df.head(10)

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   
5  BMW    1 Series  2012  premium unleaded (required)      230.0   
6  BMW    1 Series  2012  premium unleaded (required)      300.0   
7  BMW    1 Series  2012  premium unleaded (required)      300.0   
8  BMW    1 Series  2012  premium unleaded (required)      230.0   
9  BMW    1 Series  2013  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   
5               6.0            MANUAL  rear wheel drive              2.0   
6               6.0            MANUAL  rear wheel drive              2.0   
7               6.0            MANUAL  rear wheel drive              2.0   
8               6.0            MANUAL  rear wheel drive              2.0   
9               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   
5                     Luxury,Performance      Compact         Coupe   
6                     Luxury,Performance      Compact   Convertible   
7                Luxury,High-Performance      Compact         Coupe   
8                                 Luxury      Compact   Convertible   
9                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500  
5           28        18        3916  31200  
6           26        17        3916  44100  
7           28        20        3916  39300  
8           28        18        3916  36900  
9           27        18        3916  37200

In [4]:
df.shape

(11914, 16)

We see that the dataset has 11914 rows and 16 columns

### Check Top 3 Manufacturers 

In [5]:
df.groupby('Make')['Make'].count().sort_values(ascending=False).head(3)

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

### Check No. of Unique Audi Models

In [6]:
df[df['Make'] == 'Audi']['Model'].nunique()

34

### List The Unique Models

In [7]:
df[df['Make'] == 'Audi']['Model'].unique()

array(['100', '200', '80', '90', 'A3', 'A4 allroad', 'A4', 'A5', 'A6',
       'A7', 'A8', 'allroad quattro', 'allroad', 'Cabriolet', 'Coupe',
       'Q3', 'Q5', 'Q7', 'R8', 'RS 4', 'RS 5', 'RS 6', 'RS 7', 'S3', 'S4',
       'S5', 'S6', 'S7', 'S8', 'SQ5', 'TT RS', 'TT', 'TTS', 'V8'],
      dtype=object)

### Count No. of Columns Having Null Values

In [8]:
df.isnull().any().sum()

5

There are 5 columns that have null values

### Inspecting the Distribution of Null Values Per Column 

In [9]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

We see that the 'Engine Fuel Type', 'Engine HP', 'Engine Cylinders', 'Number of Doors' and 'Market Category' columns have null values and their respective counts. 

### Find Median of Engine Cylinders

In [10]:
df['Engine Cylinders'].median()

6.0

In [11]:
df['Engine Cylinders'].describe()


count    11884.000000
mean         5.628829
std          1.780559
min          0.000000
25%          4.000000
50%          6.000000
75%          6.000000
max         16.000000
Name: Engine Cylinders, dtype: float64

### Find Most Frequent Engine Cylinder Type

In [12]:
most_freq =  df['Engine Cylinders'].value_counts().sort_values(ascending=False).iloc[[0]]
most_freq

4.0    4752
Name: Engine Cylinders, dtype: int64

In [13]:
# Using index[0] to grab the index of the most frequent value
most_freq_cylinder= most_freq.index[0]
most_freq_cylinder

4.0

In [14]:
# We can also use the .mode() function which is shorter
most_freq_cylinder =  df['Engine Cylinders'].mode()
most_freq_cylinder[0]

4.0

### Fill Missing Values With Most Frequent Cylinder Value

In [15]:
df['Engine Cylinders'].fillna(most_freq_cylinder[0],inplace=True)

#### Verify if null values are filled

In [16]:
df['Engine Cylinders'].isna().any()

False

In [17]:
df['Engine Cylinders'].value_counts()

4.0     4782
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: Engine Cylinders, dtype: int64

From the output, we see that there are no null values and the no. of 4 cylinder engines has increased by 30 (which was the no. of null values that were there previously)

### Check The Median Again

In [17]:
df['Engine Cylinders'].median()

6.0

From the above, we see that the median didn't change

### Select All Lotus Cars

In [18]:
df = df[df['Make'] == 'Lotus']
df.head()

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   
3915  Exotic,High-Performance      Compact   Convertible           27   
3916  Exotic,High-Performance      Compact   Convertible           26   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250  
3915        21         613  47250  
3916        20         613  54990

In [19]:
df.shape

(29, 16)

### Select Only 'Engine HP' and 'Engine Cylinders'

In [20]:
df = df[['Engine HP', 'Engine Cylinders']]
df.head()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0

In [21]:
df.shape

(29, 2)

### Drop All Duplicate Rows

In [23]:
df.drop_duplicates(inplace=True)
df.shape

(9, 2)

### Get The Underlying Numpy Array

In [25]:
X = df.values
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

### Compute X<sup>T</sup>X

In [26]:
# Getting X transpose
XT = X.T
XT

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [27]:
# The dot function performs matrix multiplication if both inputs are matrices
XTX = XT.dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

### Invert X<sup>T</sup>X

In [30]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

### Create Array y

In [31]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

### Calculate (X<sup>T</sup>X)<sup>-1</sup>X<sup>T</sup>y

In [32]:
w = XTX_inv.dot(XT).dot(y)
w

array([  4.59494481, -63.56432501])

In [36]:
# Getting the first element of w
w[0]

4.594944810094579